packages/api-server/src/methods/validator.ts

```diff
@@ -662,6 +664,147 @@ export async function verifyEnoughBalance(
+      `gasless transaction require 0x0 gas price, got ${gasPrice}`
+    );
+  }
+
+  const toErr = verifyAddress(to, index);
+  if (toErr != null) {
+    return toErr.padContext("Gasless to address");
+  }
+
+  const gasLimitErr = verifyGasLimit(gasLimit, index);
+  if (gasLimitErr != null) {
+    return gasLimitErr.padContext("Gasless gasLimit");
+  }
+
+  const dataErr = verifyHexString(inputData, index);
+  if (dataErr != null) {
+    return dataErr.padContext("Gasless inputData");
+  }
+
+  // verify type for each field of user operation
+  const userOperation = decodeGaslessPayload(inputData);
+  // callContract
+  {
+    const err = verifyAddress(userOperation.callContract, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.callContract");
+    }
+  }
+  // callData
+  {
+    const err = verifyHexString(userOperation.callData, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.callData");
+    }
+  }
+
+  // callGasLimit
+  {
+    const err = verifyGasLimit(userOperation.callGasLimit, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.callGasLimit");
+    }
+  }
+
+  // maxFeePerGas
+  {
+    const err = verifyGasLimit(userOperation.maxFeePerGas, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.maxFeePerGas");
+    }
+  }
+
+  // maxPriorityFeePerGas
+  {
+    const err = verifyGasLimit(userOperation.maxPriorityFeePerGas, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.maxPriorityFeePerGas");
+    }
+  }
+
+  // verificationGasLimit
+  {
+    const err = verifyGasLimit(userOperation.verificationGasLimit, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.verificationGasLimit");
+    }
+  }
+
+  // paymasterAndData
+  {
+    const err = verifyHexString(userOperation.paymasterAndData, index);
+    if (err != null) {
+      return err.padContext("Gasless userOperation.paymasterAndData");
+    }
+
+    // paymasterAndData should contain paymaster address
+    if (userOperation.paymasterAndData.slice(2).length / 2 < 20) {
+      return invalidParamsError(
+        index,
+        `userOperation.paymasterAndData(${
+          userOperation.paymasterAndData.slice(2).length / 2
+        }) < 20 bytes`
+      );
+    }
+
+    const addrErr = verifyAddress(
+      userOperation.paymasterAndData.slice(0, 42),
+      index
+    );
+    if (addrErr != null) {
+      return addrErr.padContext(
+        "Gasless userOperation.paymasterAndData.address"
+      );
+    }
+  }
+
+  // extra check
+  // 1. no base fee since we are not impl eip1559
+  if (userOperation.maxFeePerGas != userOperation.maxFeePerGas) {
+    return invalidParamsError(
+      index,
+      `userOperation.maxFeePerGas(${userOperation.maxFeePerGas}) != userOperation.maxFeePerGas(${userOperation.maxFeePerGas})`
+    );
+  }
+  // 2. gasLimit = verificationGasLimit * 3 + callGasLimit
```

Comment:
Ignorant question: why should `gasLimit = verificationGasLimit * 3 + callGasLimit`? Specifically, why `* 3`?

Reply:
https://github.com/godwokenrises/account-abstraction/blob/gw-gasless/contracts/core/GaslessEntryPoint.sol#L327-L338

similar to https://github.com/godwokenrises/account-abstraction/blob/develop/contracts/core/EntryPoint.sol#L265-L268